Doxygen/Docs/man/man3/Camera.3
.TH "Camera" 3 "Fri Mar 29 2013" "Version 31337" "HyperGrafx" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Camera \- 
.PP
The \fBCamera\fP class represents a logical camera in a model view, which posesses a current viewing angle and an absolute position in space as its state\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Camera\&.hpp>\fP
.PP
Inherits \fBObject\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBDirection\fP { \fBDIR_FORWARD\fP, \fBDIR_BACKWARD\fP, \fBDIR_LEFT\fP, \fBDIR_RIGHT\fP, \fBDIR_UP\fP, \fBDIR_DOWN\fP, \fBDIR_END\fP, \fBDIR_BEGIN\fP =  DIR_FORWARD }"
.br
.RI "\fIThe Direction enumeration lists all of the possible directions the camera may travel in\&. \fP"
.ti -1c
.RI "enum \fBViewType\fP { \fBPERSPECTIVE\fP, \fBORTHO\fP, \fBORTHO2D\fP, \fBIDENTITY\fP, \fBFRUSTUM\fP }"
.br
.RI "\fIThe ViewType enumeration lists the various possibilities for the current viewing mode that can be switched between\&. \fP"
.ti -1c
.RI "enum \fBUniforms\fP { \fBBEGIN\fP =  Object::END, \fBTRANSLATION\fP =  BEGIN, \fBROTATION\fP, \fBVIEW\fP, \fBCTM\fP, \fBEND\fP }"
.br
.RI "\fIThe glsl_var enumeration lists the various variables the \fBCamera\fP class is capable of sending to the shader\&. \fP"
.ti -1c
.RI "typedef enum \fBCamera::Direction\fP \fBDirection\fP"
.br
.RI "\fIThe Direction enumeration lists all of the possible directions the camera may travel in\&. \fP"
.ti -1c
.RI "typedef enum \fBCamera::ViewType\fP \fBViewType\fP"
.br
.RI "\fIThe ViewType enumeration lists the various possibilities for the current viewing mode that can be switched between\&. \fP"
.ti -1c
.RI "typedef enum \fBCamera::Uniforms\fP \fBUniform\fP"
.br
.RI "\fIThe glsl_var enumeration lists the various variables the \fBCamera\fP class is capable of sending to the shader\&. \fP"
.ti -1c
.RI "typedef const unsigned int \fBUniformEnum\fP"
.br
.RI "\fIThe \fBObject\fP class takes advantage of child-extendible enumerations\&. \fP"
.ti -1c
.RI "typedef std::map
.br
< \fBObject::UniformEnum\fP, 
.br
std::string > \fBUniformMap\fP"
.br
.RI "\fIWe store mappings of Uniform Enumerations, The desired function of the var, to strings, the names of the variables\&. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCamera\fP (const std::string &\fBname\fP, GLuint gShader, float \fBx\fP=0\&.0, float \fBy\fP=0\&.0, float \fBz\fP=0\&.0)"
.br
.RI "\fIInitialization Constructor; sets the x,y,z coordinates explicitly\&. \fP"
.ti -1c
.RI "\fBCamera\fP (const std::string &\fBname\fP, GLuint gShader, \fBvec3\fP &in)"
.br
.RI "\fIInitialization Constructor, uses a vec3 as its initial coordinates\&. \fP"
.ti -1c
.RI "\fBCamera\fP (const std::string &\fBname\fP, GLuint gShader, \fBvec4\fP &in)"
.br
.RI "\fIInitialization Constructor, uses a vec4 as its initial coordinates\&. \fP"
.ti -1c
.RI "virtual \fB~Camera\fP (void)"
.br
.RI "\fIDefault destructor\&. \fP"
.ti -1c
.RI "void \fBx\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the x coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBy\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the y coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBz\fP (const float &in, const bool &update=true)"
.br
.RI "\fISets the z coordinate of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const float &\fBx\fP, const float &\fBy\fP, const float &\fBz\fP, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const \fBvec3\fP &in, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBpos\fP (const \fBvec4\fP &in, const bool &update=true)"
.br
.RI "\fISets the absolute position of the camera\&. \fP"
.ti -1c
.RI "void \fBdX\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the x axis\&. \fP"
.ti -1c
.RI "void \fBdY\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the y axis\&. \fP"
.ti -1c
.RI "void \fBdZ\fP (const float &by, const bool &update=true)"
.br
.RI "\fIMoves the camera along the z axis\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const float &\fBx\fP, const float &\fBy\fP, const float &\fBz\fP)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const \fBvec3\fP &by)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBdPos\fP (const \fBvec4\fP &by)"
.br
.RI "\fIMoves the camera along the x, y, and z axes\&. \fP"
.ti -1c
.RI "void \fBfieldOfView\fP (const float &fovy)"
.br
.RI "\fIfieldOfView sets the current camera Field-of-view angle\&. \fP"
.ti -1c
.RI "float \fBfieldOfView\fP (void) const "
.br
.RI "\fI\fBfieldOfView()\fP gets the current camera Field-of-view angle\&. \fP"
.ti -1c
.RI "void \fBadjustFieldOfView\fP (const float &by)"
.br
.RI "\fIadjustFieldOfView adjusts the field of view angle up or down by an amount\&. \fP"
.ti -1c
.RI "void \fBchangePerspective\fP (const \fBViewType\fP &vType)"
.br
.RI "\fIchangePerspective changes the current perspective of the camera\&. \fP"
.ti -1c
.RI "void \fBrefreshPerspective\fP (void)"
.br
.RI "\fIrefreshPerspective re-generates the current view/perspective matrix of the camera\&. \fP"
.ti -1c
.RI "void \fBviewport\fP (size_t _X, size_t _Y, size_t _width, size_t _height)"
.br
.RI "\fIviewport instructs this camera what his expected drawing window will be\&. \fP"
.ti -1c
.RI "void \fBsway\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's x coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBsurge\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's z coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBheave\fP (const float &by)"
.br
.RI "\fIAdjusts the camera's y coordinate relative to its current position\&. \fP"
.ti -1c
.RI "void \fBpitch\fP (const float &by, const bool &fixed=false)"
.br
.RI "\fIpitch adjusts the x axis rotation; up/down look\&. \fP"
.ti -1c
.RI "void \fByaw\fP (const float &by, const bool &fixed=false)"
.br
.RI "\fIyaw adjusts the y axis rotation; left/right look\&. \fP"
.ti -1c
.RI "void \fBroll\fP (const float &by, const bool &fixed=false)"
.br
.RI "\fIroll adjusts the z axis rotation; tilt or lean left/right\&. \fP"
.ti -1c
.RI "void \fBmove\fP (const \fBCamera::Direction\fP &Dir)"
.br
.RI "\fImove instructs the camera to begin moving in the specified direction\&. \fP"
.ti -1c
.RI "void \fBstop\fP (const \fBCamera::Direction\fP &Dir)"
.br
.RI "\fIstop instructs the camera to stop moving in the specified direction\&. \fP"
.ti -1c
.RI "void \fBidle\fP (void)"
.br
.RI "\fIidle moves the camera forward in whichever directions it is configured to move in\&. \fP"
.ti -1c
.RI "void \fBaccel\fP (const \fBvec3\fP &accel)"
.br
.RI "\fIaccel takes an input vec2 which represents an acceleration, and applies it to the motion vectors with regards to the maximum acceleration and the maximum speed of the camera\&. \fP"
.ti -1c
.RI "float \fBx\fP (void) const "
.br
.RI "\fI\fBx()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "float \fBy\fP (void) const "
.br
.RI "\fI\fBy()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "float \fBz\fP (void) const "
.br
.RI "\fI\fBz()\fP returns the current position of the camera in model coordinates\&. \fP"
.ti -1c
.RI "\fBvec4\fP \fBpos\fP (void) const "
.br
.RI "\fI\fBpos()\fP gets the current camera position in model coordinates\&. \fP"
.ti -1c
.RI "virtual void \fBsend\fP (\fBObject::UniformEnum\fP which)"
.br
.RI "\fIsend will send a glsl variable to the shader\&. \fP"
.ti -1c
.RI "void \fBview\fP (void)"
.br
.RI "\fIview will instruct OpenGL of the viewport we want, and then send all of our current matrices to the shader for rendering\&. \fP"
.ti -1c
.RI "void \fBresetRotation\fP (void)"
.br
.RI "\fIresetRotation adjusts the camera's rotational state back to its default state (The Identity Matrix\&.) \fP"
.ti -1c
.RI "void \fBdraw\fP (void)"
.br
.RI "\fIdraw method: Render this object to the screen _buffer\&. \fP"
.ti -1c
.RI "void \fBbuffer\fP (void)"
.br
.RI "\fIbuffer all of our data: Vertices, TexUVs, Normals, Indices, Colors and Morph Buffers\&. \fP"
.ti -1c
.RI "void \fBbufferMorphOnly\fP (void)"
.br
.RI "\fIbuffer only the Morph-related buffers\&. \fP"
.ti -1c
.RI "void \fBdrawMode\fP (GLenum new_mode)"
.br
.RI "\fISelect a new OpenGL draw mode for this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBtexture\fP (const char **filename)"
.br
.RI "\fIFIXME: This is a junk, nonflexible method\&. \fP"
.ti -1c
.RI "const std::string & \fBname\fP (void) const "
.br
.RI "\fIRetrieve the _name of this \fBObject\fP\&. \fP"
.ti -1c
.RI "virtual void \fBlink\fP (\fBUniformEnum\fP which, const std::string &\fBname\fP)"
.br
.RI "\fIlink a specified Uniform against the shader's variable _name\&. \fP"
.ti -1c
.RI "virtual GLuint \fBshader\fP (void)"
.br
.RI "\fIReturns the \fBObject\fP's current shader\&. \fP"
.ti -1c
.RI "virtual void \fBshader\fP (GLuint newShader)"
.br
.RI "\fISets the shader to be used by this object\&. \fP"
.ti -1c
.RI "void \fBanimation\fP (void(*anim_func)(\fBTransCache\fP &arg))"
.br
.RI "\fIApply an animation callback function to this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBpropagate\fP (void)"
.br
.RI "\fIScene-graph changes are not automatically applied to children\&. \fP"
.ti -1c
.RI "\fBvec4\fP \fBposition\fP () const "
.br
.RI "\fIObtain the vec4 representative of the \fBObject\fP's current position in space\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBmorphTarget\fP () const "
.br
.RI "\fIRetrieve a pointer to this object's morph target\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBgenMorphTarget\fP (GLuint \fBshader\fP)"
.br
.RI "\fIInstantiate a new morphing target\&. \fP"
.ti -1c
.RI "float \fBmorphPercentage\fP () const "
.br
.RI "\fIRetrieve the morph Percentage of this object\&. \fP"
.ti -1c
.RI "void \fBmorphPercentage\fP (const float newPercentage)"
.br
.RI "\fISet the morph percentage of this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBdestroyMorphTarget\fP ()"
.br
.RI "\fIObliterate the morph target for this object\&. \fP"
.ti -1c
.RI "int \fBnumberOfPoints\fP ()"
.br
.RI "\fIRetrieve the number of _vertices this object has\&. \fP"
.ti -1c
.RI "void \fBtextureID\fP (GLint newTextureID)"
.br
.RI "\fISet the \fBTexture\fP ID / \fBTexture\fP Unit for this \fBObject\fP\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBaddObject\fP (const std::string &objName, GLuint Object_Shader=0)"
.br
.RI "\fIaddObject creates a new \fBObject\fP with the given name and, optionally, a specified shader and adds it to the \fBScene\fP graph\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (const std::string &objName)"
.br
.RI "\fIdelObject will remove from the \fBScene\fP graph the object with the given name\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (void)"
.br
.RI "\fIdelObject with no parameters will delete the first \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "void \fBpopObject\fP (void)"
.br
.RI "\fIpopObject deletes the last \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBnext\fP (void)"
.br
.ti -1c
.RI "\fBObject\fP * \fBprev\fP (void)"
.br
.ti -1c
.RI "\fBObject\fP * \fBactive\fP (void) const "
.br
.ti -1c
.RI "\fBObject\fP * \fBoperator[]\fP (const std::string &objname)"
.br
.ti -1c
.RI "void \fBinsertObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIRegister a created object with the scene graph\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::vector< \fBAngel::vec4\fP > \fB_vertices\fP"
.br
.RI "\fIvertex buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec3\fP > \fB_normals\fP"
.br
.RI "\fINormals buffer\&. \fP"
.ti -1c
.RI "std::vector< unsigned int > \fB_indices\fP"
.br
.RI "\fIDraw Order Index buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec4\fP > \fB_colors\fP"
.br
.RI "\fIColors buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec2\fP > \fB_texUVs\fP"
.br
.RI "\fI\fBTexture\fP Coordinates buffer\&. \fP"
.ti -1c
.RI "\fBTransCache\fP \fB_trans\fP"
.br
.RI "\fIThe _trans cache encompasses the current transformational state of this object\&. \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBdeleteObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIVery seriously delete a child object and free his memory\&. \fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "std::string \fB_name\fP"
.br
.RI "\fI_name is used as an identifying handle for the object\&. \fP"
.ti -1c
.RI "GLuint \fB_vao\fP"
.br
.RI "\fIVertex Array \fBObject\fP handle identifying our buffers/object\&. \fP"
.ti -1c
.RI "GLuint \fB_buffer\fP [\fBNUM_BUFFERS\fP]"
.br
.RI "\fIHandles to our buffers (Vertices, TexUVs, etc\&.) \fP"
.ti -1c
.RI "GLenum \fB_drawMode\fP"
.br
.RI "\fIDrawing mode for this object\&. \fP"
.ti -1c
.RI "bool \fB_isTextured\fP"
.br
.RI "\fIIs this object textured? \fP"
.ti -1c
.RI "float \fB_morphPercentage\fP"
.br
.RI "\fIThe percentage of the morph\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fB_morphTarget\fP"
.br
.RI "\fIA pointer to the object we wish to morph into\&. \fP"
.ti -1c
.RI "std::map< \fBObject::UniformEnum\fP, 
.br
std::string > \fB_uniformMap\fP"
.br
.RI "\fIA map between Uniform variable functions and the actual uniform variable names\&. \fP"
.ti -1c
.RI "std::vector< GLint > \fB_handles\fP"
.br
.RI "\fIHandles to Uniforms on the shader\&. \fP"
.ti -1c
.RI "GLint \fB_textureID\fP"
.br
.RI "\fIThe texture unit index this \fBObject\fP uses\&. \fP"
.ti -1c
.RI "std::list< \fBObject\fP * > \fB_list\fP"
.br
.RI "\fIFor the purposes of rapid propagation of scene-graph changes, \fBObject\fP pointers are stored in a regular flat list\&. \fP"
.ti -1c
.RI "std::map< std::string, \fBObject\fP * > \fB_map\fP"
.br
.RI "\fIFor the purposes of accessing named objects quickly, though, objects are also re-stored in an associative map\&. \fP"
.ti -1c
.RI "std::list< \fBObject\fP * >::iterator \fB_currentObj\fP"
.br
.RI "\fIWe keep an iterator on-hand that references what the scene considers to be it's active, current object\&. \fP"
.ti -1c
.RI "GLuint \fB_gShader\fP"
.br
.RI "\fIA handle to a shader program to be used as the default shader for new children objects added to the scene\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBadjustRotation\fP (const \fBmat4\fP &adjustment, const bool &fixed=false)"
.br
.RI "\fIadjustRotation is an internal function that rotates the camera\&. \fP"
.ti -1c
.RI "void \fBcommonInit\fP (void)"
.br
.RI "\fIcommonInit is a private function that initializes local object attributes\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBmat4\fP \fB_view\fP"
.br
.RI "\fIThe current view matrix (defaultly perspective) for this camera\&. \fP"
.ti -1c
.RI "\fBTransCache\fP \fB_ctm\fP"
.br
.RI "\fIThe Current \fBTransformation\fP state for this \fBCamera\fP\&. \fP"
.ti -1c
.RI "\fBViewType\fP \fB_currentView\fP"
.br
.RI "\fIThe current viewing mode type\&. \fP"
.ti -1c
.RI "GLfloat \fB_speed\fP"
.br
.RI "\fICurrent Speed of camera motion\&. \fP"
.ti -1c
.RI "\fBvec3\fP \fB_velocity\fP"
.br
.RI "\fICurrent Velocity of camera motion\&. \fP"
.ti -1c
.RI "GLfloat \fB_speed_cap\fP"
.br
.RI "\fICurrent Speed Capacity: (speed/MaxSpeed) \fP"
.ti -1c
.RI "GLfloat \fB_maxAccel\fP"
.br
.RI "\fIMaximum Acceleration Magnitude\&. \fP"
.ti -1c
.RI "GLfloat \fB_maxSpeed\fP"
.br
.RI "\fIMaximum Speed\&. \fP"
.ti -1c
.RI "GLfloat \fB_frictionMagnitude\fP"
.br
.RI "\fIFriction\&. \fP"
.ti -1c
.RI "GLfloat \fB_aspectRatio\fP"
.br
.RI "\fICurrent aspect ratio for certain perspectives\&. \fP"
.ti -1c
.RI "GLfloat \fB_fovy\fP"
.br
.RI "\fICurrent field-of-view angle for perspective view\&. \fP"
.ti -1c
.RI "\fBAngel::vec2\fP \fB_viewportSize\fP"
.br
.RI "\fI\fBCamera\fP's Drawbox Width and Height\&. \fP"
.ti -1c
.RI "\fBAngel::vec2\fP \fB_viewportPosition\fP"
.br
.RI "\fI\fBCamera\fP's Drawbox x,y Coordinate (Upper-Left Pixel) \fP"
.ti -1c
.RI "bool \fB_motion\fP [Camera::DIR_END]"
.br
.RI "\fIBooleans correlating to the different motion directions\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
The \fBCamera\fP class represents a logical camera in a model view, which posesses a current viewing angle and an absolute position in space as its state\&. 

\fBAuthor:\fP
.RS 4

GraphicsProject/Camera.cpp
/**
 @file Camera.cpp
 @author John Huston
 @authors John Huston, Nicholas StPierre, Chris Compton
 @date 2012-12-20
 @brief Implementation for the Camera class.
 **/

#include <stdexcept>
#include <iostream>

#include "mat.hpp"
#include "vec.hpp"
#include "Camera.hpp"
#include "globals.h" //Math constants and macros (SQRT2, POW5)
#include "Timer.hpp" //Global timer to sync with framerate.
using namespace Angel;

void Camera::commonInit( void ) {

  // Extend the Uniforms array.
  if ( DEBUG )
    fprintf( stderr, "Extending Uniforms Array to %d\n", Camera::END );
  this->_handles.resize( Camera::END, -1 );

  /* Default Variable Links */
  link( Camera::TRANSLATION, "T" );
  link( Camera::ROTATION, "R" );
  link( Camera::VIEW, "P" );
  link( Camera::CTM, "CTM" );

  for ( size_t i = (size_t) Camera::DIR_BEGIN; i != (size_t) DIR_END; ++i )
    _motion[i] = false;

  this->_speed = 0;
  this->_speed_cap = 0;
  this->_maxAccel = 10;
  this->_maxSpeed = 2000;
  this->_frictionMagnitude = 4;
  this->_aspectRatio = 1;
  this->_currentView = PERSPECTIVE;
  this->_fovy = 45.0;
}

Camera::Camera( const std::string &name, GLuint gShader, float x, float y,
                float z ) :
    Object( name, gShader ) {
  commonInit();
  this->pos( x, y, z, false );
}

Camera::Camera( const std::string &name, GLuint gShader, vec3 &in ) :
    Object( name, gShader ) {
  commonInit();
  this->pos( in, false );
}

Camera::Camera( const std::string &name, GLuint gShader, vec4 &in ) :
    Object( name, gShader ) {
  commonInit();
  this->pos( in, false );
}

Camera::~Camera( void ) {

}

void Camera::x( const float &in, const bool &update ) {

  _ctm.offset.SetX( -in );
  _trans.offset.SetX( in );
  if ( update ) {
    _ctm.CalcCTM();
    _trans.CalcCTM();
    send( TRANSLATION );
  }
}

void Camera::y( const float &in, const bool &update ) {
  _ctm.offset.SetY( -in );
  _trans.offset.SetY( in );
  if ( update ) {
    _ctm.CalcCTM();
    _trans.CalcCTM();
    send( TRANSLATION );
  }
}

void Camera::z( const float &in, const bool &update ) {
  _ctm.offset.SetZ( -in );
  _trans.offset.SetZ( in );
  if ( update ) {
    _trans.CalcCTM();
    _ctm.CalcCTM();
    send( TRANSLATION );
  }
}

void Camera::pos( const float &newX, const float &newY, const float &newZ,
                  const bool &update ) {
  x( newX, false );
  y( newY, false );
  z( newZ, false );
  if ( update ) {
    _trans.CalcCTM();
    _ctm.CalcCTM();
    send( TRANSLATION );
  }
}

void Camera::pos( const vec4 &in, const bool &update ) {
  this->pos( in.x, in.y, in.z, update );
}

void Camera::pos( const vec3 &in, const bool &update ) {
  this->pos( in.x, in.y, in.z, update );
}

void Camera::dX( const float &by, const bool &update ) {
  x( x() + by, update );
}

void Camera::dY( const float &by, const bool &update ) {
  y( y() + by, update );
}

void Camera::dZ( const float &by, const bool &update ) {
  z( z() + by, update );
}

void Camera::dPos( const float &x, const float &y, const float &z ) {
  dX( x, false );
  dY( y, false );
  dZ( z, false );
  _trans.CalcCTM();
  _ctm.CalcCTM();
  send( TRANSLATION );
}

void Camera::dPos( const vec3 &by ) {
  dPos( by.x, by.y, by.z );
}

void Camera::dPos( const vec4 &by ) {
  dPos( by.x, by.y, by.z );
}

void Camera::adjustRotation( const mat4 &adjustment, const bool &fixed ) {

  /*
   By default, the 'order' bool represents the PREMULT behavior.
   However, if the fixed bool is present, toggle the order bool.
   This stealthily inserts this rotation "first",
   So that it is applied before any other rotations.
   */
  bool order = POSTMULT;
  if ( fixed ) order = !order;

  // Apply our rotational adjustment to the camera.
  // Unintuitively, we need to adjust the Orbit.
  _ctm.orbit.Adjust( adjustment, order );

  /*
   Next, our Camera may have a physical object whose
   Rotations need to be calculated as well.
   We need to apply the OPPOSITE rotations,
   So Transpose the Adjustment to obtain that.
   */
  _trans.rotation.Adjust( transpose( adjustment ), !order );

  // Update our state.
  _trans.CalcCTM();
  _ctm.CalcCTM();
  send( ROTATION );
}

/**
 ROTATE_OFFSET is a macro which is used to normalize
 the six camera motion directions with respect to the
 current camera rotation. It is used in heave(), sway() and surge().
 @param V a vec4 representing the movement offset vector.
 @return A rotated vec4.
 **/
#define ROTATE_OFFSET(V) (V * _ctm.orbit.Matrix())

void Camera::sway( const float &by ) {
  dPos( ROTATE_OFFSET(vec4(by,0,0,0)) );
}

void Camera::surge( const float &by ) {
  dPos( ROTATE_OFFSET(vec4(0,0,-by,0)) );
}

void Camera::heave( const float &by ) {
  dPos( ROTATE_OFFSET(vec4(0,by,0,0)) );
}

void Camera::pitch( const float &by, const bool &fixed ) {
  /*
   Since negative values are interpreted as pitching down,
   We leave the input uninverted, because a negative rotation
   about the x axis rotates counter-clockwise (looking right),
   and clockwise (looking left), which achieves the effect of
   looking 'down'.
   */
  adjustRotation( RotateX( -by ), fixed );
}

void Camera::yaw( const float &by, const bool &fixed ) {
  /*
   Since a positive 'by' should represent looking right,
   we invert the rotation because rotating by a positive value
   will rotate right, which simulates looking left.
   Therefore, invert.
   */
  adjustRotation( RotateY( by ), fixed );
}

void Camera::roll( const float &by, const bool &fixed ) {
  adjustRotation( RotateZ( by ), fixed );
}

void Camera::accel( const vec3 &raw_accel ) {

  /*
   This scale factor is the cumulation of several scaling factors.
   (A) (MaxAccel/SQRT3)
   scales the vector so that it has a magnitude
   Between -MAX_ACCEL and MAX_ACCEL.

   (B) (1-POW5(speed_cap))
   scales the vector with relation to the
   current speed. This effectively models "Drag".
   The closer we are to maximum speed, this scaling factor
   Will approach 0, so that if we are at MAX SPEED,
   our acceleration will be scaled to (nearly) zero.

   (C) (Tick.Scale())
   This scales the amount of acceleration by an amount [0,1]
   so that if we are drawing many frames, we'll apply less acceleration.
   If we are drawing not so many, we'll apply more.
   This should keep the animation relatively consistent across platforms.
   */

  float Scale = (_maxAccel / SQRT3) * (1 - POW5(_speed_cap)) * (Tick.Scale());
  vec3 accel = raw_accel * Scale;

  if ( DEBUG_MOTION ) {
    fprintf( stderr, "Accel(); raw_accel = (%f,%f,%f)\n", raw_accel.x,
             raw_accel.y, raw_accel.z );
    fprintf(
        stderr,
        "Accel(); Scale = (MaxAccel/SQRT3) * (1-POW5(speed_cap)) * Tick.Scale()\n" );
    fprintf( stderr, "Accel(); Scale = (%f/%f) * (%f) * (%f)\n", _maxAccel,
             SQRT3, (1 - POW5(_speed_cap)), Tick.Scale() );
    fprintf( stderr, "Accel(); Scale = %f\n", Scale );
    fprintf( stderr, "Accle(); accel = raw_accel * Scale = (%f,%f,%f)\n",
             accel.x, accel.y, accel.z );
  }

  //The acceleration is finally applied to the velocity vector.
  _velocity += accel;

  //speed and speed_cap must now be recalculated.
  _speed_cap = (_speed = length( _velocity )) / _maxSpeed;

  if ( DEBUG_MOTION )
    fprintf( stderr, "Applied Acceleration to Velocity, Is now: (%f,%f,%f)\n",
             _velocity.x, _velocity.y, _velocity.z );
}

void Camera::move( const Camera::Direction &Dir ) {
  _motion[Dir] = true;
}

void Camera::stop( const Camera::Direction &Dir ) {
  _motion[Dir] = false;
}

void Camera::idle( void ) {

  /* Apply the automated motion instructions, if any --
   These are primarily from the keyboard. */
  if ( _motion[Camera::DIR_FORWARD] ) accel( vec3( 0, 0, 1 ) );
  if ( _motion[Camera::DIR_BACKWARD] ) accel( vec3( 0, 0, -1 ) );
  if ( _motion[Camera::DIR_RIGHT] ) accel( vec3( 0, 1, 0 ) );
  if ( _motion[Camera::DIR_LEFT] ) accel( vec3( 0, -1, 0 ) );
  if ( _motion[Camera::DIR_UP] ) accel( vec3( 1, 0, 0 ) );
  if ( _motion[Camera::DIR_DOWN] ) accel( vec3( -1, 0, 0 ) );

  if ( _speed ) {
    /* Apply the velocity vectors computed from accel,
     which includes instructions from keyboard and the Balance Board. */
    /* 1/20000 is a magic constant which converts our velocity units
     into model units. */
    /* Tick.Scale() helps keep animation speed consistent
     between different hardware. */
    float UnitScale = (1.0 / 20000.0);
    float Scale = Tick.Scale() * UnitScale;

    if ( DEBUG_MOTION )
      fprintf( stderr, "Applying Translation: + (%f,%f,%f)\n",
               _velocity.x * Scale, _velocity.y * Scale, _velocity.z * Scale );

    heave( _velocity.x * Scale );
    sway( _velocity.y * Scale );
    surge( _velocity.z * Scale );

    // Friction Calculations
    if ( _speed < (_frictionMagnitude * Tick.Scale()) ) {
      if ( DEBUG_MOTION )
        fprintf( stderr, "Friction has stopped all movement.\n" );
      _velocity = vec3( 0, 0, 0 );
      _speed = 0;
      _speed_cap = 0;
    } else {
      // Friction is a vector that is the opposite of velocity.
      vec3 frictionVec = -_velocity;
      /* By dividing friction by (speed/FrictionMagnitude),
       we guarantee that the magnitude is FrictionMagnitude. */
      frictionVec = frictionVec / (_speed / _frictionMagnitude);
      frictionVec *= Tick.Scale();

      if ( DEBUG_MOTION )
        fprintf( stderr, "Applying friction to Velocity: + (%f,%f,%f)\n",
                 frictionVec.x, frictionVec.y, frictionVec.z );
      _velocity += frictionVec;
      _speed = length( _velocity );
      _speed_cap = _speed / _maxSpeed;
    }
  }
}

float Camera::x( void ) const {
  return -_ctm.offset.Matrix()[0][3];
}

float Camera::y( void ) const {
  return -_ctm.offset.Matrix()[1][3];
}

float Camera::z( void ) const {
  return -_ctm.offset.Matrix()[2][3];
}

vec4 Camera::pos( void ) const {
  return vec4( x(), y(), z(), 1.0 );
}

float Camera::fieldOfView( void ) const {
  return _fovy;
}

void Camera::fieldOfView( const float &in ) {
  _fovy = in;
  if ( _currentView == Camera::PERSPECTIVE ) refreshPerspective();
}

void Camera::changePerspective( const ViewType &vType ) {

  _currentView = vType;
  refreshPerspective();

}

void Camera::refreshPerspective( void ) {

  // Some constants. For your pleasure.
  static const GLfloat zNear = 0.001;
  static const GLfloat zFar = 100.0;

  switch ( _currentView ) {
  case PERSPECTIVE:
    _view = Perspective( _fovy, _aspectRatio, zNear, zFar );
    break;
  case ORTHO:
    _view = Ortho( -1.0, 1.0, -1.0, 1.0, zNear, zFar );
    break;
  case ORTHO2D:
    _view = Ortho2D( -1.0, 1.0, -1.0, 1.0 );
    break;
  case FRUSTUM:
    _view = Frustum( -1.0, 1.0, -1.0, 1.0, zNear, zFar );
    break;
  case IDENTITY:
  default:
    _view = mat4( GLuint( 1.0 ) );
    break;
  }
}

void Camera::adjustFieldOfView( const float &by ) {
  fieldOfView( fieldOfView() + by );
}

void Camera::viewport( size_t _X, size_t _Y, size_t _Width, size_t _Height ) {
  this->_viewportPosition = Angel::vec2( _X, _Y );
  this->_viewportSize = Angel::vec2( _Width, _Height );
  this->_aspectRatio = (double) _Width / (double) _Height;
  refreshPerspective();
}

void Camera::send( Object::UniformEnum which ) {

  switch ( which ) {
  case TRANSLATION:
    if ( _handles[which] != -1 )
    glUniformMatrix4fv( _handles[which], 1, GL_TRUE, _ctm.offset.Matrix() );
    send( CTM );
    break;
  case ROTATION:
    if ( _handles[which] != -1 )
    glUniformMatrix4fv( _handles[which], 1, GL_TRUE, _ctm.orbit.Matrix() );
    send( CTM );
    break;
  case VIEW:
    if ( _handles[which] != -1 )
    glUniformMatrix4fv( _handles[which], 1, GL_TRUE, _view );
    send( CTM );
    break;
  case CTM:
    _ctm.CalcCTM( POSTMULT );
    if ( _handles[which] != -1 )
    glUniformMatrix4fv( _handles[which], 1, GL_TRUE, _ctm.OTM() );
    break;
  default:
    // If we don't know which variable this is,
    // See if our parent method knows.
    Object::send( which );
    break;
  }
}

void Camera::view( void ) {

  glViewport( _viewportPosition.x, _viewportPosition.y, _viewportSize.x,
              _viewportSize.y );
  /* send all of our matrices, who knows what the shader's gonna do with 'em */
  send( TRANSLATION );
  send( ROTATION );
  send( VIEW );
  send( CTM );

}

void Camera::resetRotation( void ) {

  // The transpose of any rotation is its inverse.
  // Thus, this resets the rotational matrix.
  this->_ctm.orbit.Adjust( transpose( this->_ctm.rotation.Matrix() ) );

}
